Fix pod start duration observation

ali-a-a opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Try to fix issue #132268. It seems that the logic for observing pod startup duration is incorrect. When the pod status is not present in the status manager, kubelet records the duration (which doesn't seem right). Later, when the pod is actually running and it is in the status manager, it observes the pod startup again. I suppose it is the reason kubelet counts pod startups twice.

I added the log below for debugging, and it is the output before the fix.

```go
existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
if !ok || existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
	!firstSeenTime.IsZero() {
        // Debug log
	klog.V(0).Infof("Observe pod start duration: pod.UID: %v, ok: %v, existingStatus.Phase: %v, apiPodStatus.Phase: %v, firstSeen: %v", pod.UID, ok, existingStatus.Phase, apiPodStatus.Phase, firstSeenTime)
	metrics.PodStartDuration.Observe(metrics.SinceInSeconds(firstSeenTime))
}
```

```bash
Observe pod start duration: pod.UID: 009e1a2b-4f75-4b4d-b30e-fd03ad52caad, ok: false, existingStatus.Phase: , apiPodStatus.Phase: Pending, firstSeen: 2025-06-13 03:57:50.571995801 +0000 UTC
Observe pod start duration: pod.UID: 009e1a2b-4f75-4b4d-b30e-fd03ad52caad, ok: true, existingStatus.Phase: Pending, apiPodStatus.Phase: Running, firstSeen: 2025-06-13 03:57:50.571995801 +0000 UTC
```

The pod in the first log is not running yet, but we observe its startup incorrectly. After the fix, the log happens only once - when the pod is actually running.


#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Fixes #132268

#### Special notes for your reviewer:

In addition to the above test, I verified that the `kubelet_pod_start_duration_seconds` metric increments its count by one with each pod creation after applying the fix.

My Kubelet is compiled using the master branch.  Also, my Containerd's version is as follows:

```bash
containerd github.com/containerd/containerd/v2 v2.1.1 cb1076646aa3740577fafbf3d914198b7fe8e3f7
```

Before the fix:

```bash
curl -k -H "Authorization: Bearer $METRICS_READER_TOKEN" https://localhost:10250/metrics | grep kubelet_pod_start_duration_seconds_count
kubelet_pod_start_duration_seconds_count 15

kubectl run nginx --image=nginx 

curl -k -H "Authorization: Bearer $METRICS_READER_TOKEN" https://localhost:10250/metrics | grep kubelet_pod_start_duration_seconds_count
kubelet_pod_start_duration_seconds_count 17
```

After the fix:

```bash
curl -k -H "Authorization: Bearer $METRICS_READER_TOKEN" https://localhost:10250/metrics | grep kubelet_pod_start_duration_seconds_count
kubelet_pod_start_duration_seconds_count 3

kubectl run nginx --image=nginx

curl -k -H "Authorization: Bearer $METRICS_READER_TOKEN" https://localhost:10250/metrics | grep kubelet_pod_start_duration_seconds_count
kubelet_pod_start_duration_seconds_count 4
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix the issue with the `kubelet_pod_start_duration_seconds` metric being recorded twice per pod.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ali-a-a committed:
Fix pod start duration observation

natasha41575 reviewed (commented):
@HirazawaUi I actually was able to reproduce the issue after trying a few more times - left a comment on https://github.com/kubernetes/kubernetes/issues/132268.

ali-a-a commented on pkg/kubelet/kubelet.go:
```diff
@@ -1928,7 +1928,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 	// since kubelet first saw the pod if firstSeenTime is set.
 	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
-	if !ok || existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
+	if ok && existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
```
> I think this the preferable way to fix the issue. With the change this PR has now, it would not record the metric until the second sync loop iteration, which is less ideal.

Hmm. After looking at the statusManager again, I think your suggested code is the intended logic. There can be a case that the pod isn't present in the manager, but it is running. My fix wouldn't handle this case.

I'll push the corrected version soon.

HirazawaUi reviewed (commented):
/hold

Thanks for your quick action, but this is a bug fix. I think we should add an e2e test to ensure that the issue is indeed resolved.

natasha41575 commented:
> I think we should add an e2e test to ensure that the issue is indeed resolved.

agreed that test coverage is helpful - however a full e2e test is probably not needed, a unit test in kubelet_test.go should be sufficient

ali-a-a commented:
> I think we should add an e2e test to ensure that the issue is indeed resolved.

@HirazawaUi I also agree that we should test the bug fix, but an e2e test for a single metric seems a bit too much. If you are all ok with it, I can add a unit test for the `kubelet_pod_start_duration_seconds` metric.

HirazawaUi commented:
> I also agree that we should test the bug fix, but an e2e test for a single metric seems a bit too much. If you are all ok with it, I can add a unit test for the `kubelet_pod_start_duration_seconds` metric.

I don't have a strong preference here, but typically an e2e test provides more visual verification of whether a bug is actually fixed. If you believe a unit test is sufficient, I'd suggest you:
  * Verify the fix through local testing
  * Document your test results in the PR description (While I've confirmed locally that this fix works well), this would significantly help approvers during review.

ali-a-a committed:
Add TestPodStartDurationCountMetric

ali-a-a commented:
@aojea @HirazawaUi 

I added a unit test that fails without the fix. I also added a short document in my PR's description on how I tested the cluster before and after the fix.

About e2e tests, I couldn't find anything related to Kubelet's metrics (other than the instrumentation package which only grabs all the metrics). Still, let me know if you think the unit test is not sufficient and it is better to add an e2e test as well.

ali-a-a commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -1407,6 +1407,70 @@ func TestCreateMirrorPod(t *testing.T) {
+				{Name: "containerA"},
+			},
+		},
+		Status: v1.PodStatus{
```
As you mentioned, we already cover these cases: 

- `ok` is false and `apiPodStatus.Phase == v1.PodPending`
- `ok` is true, `existingStatus.Phase == v1.PodPending`,  and `apiPodStatus.Phase == v1.PodRunning`

I think the only case that remains is when `existingStatus.Phase == v1.PodRunning`, which means that we should have already observed the pod startup, and duplication shouldn't happen in the metric count. We can add this case by simply adding another `SyncPod` to ensure that it won't observe the pod again. The added `SyncPod` call would be something like this:

```go
// Syncing the running pod for the second time.
_, err = kubelet.SyncPod(context.Background(), kubetypes.SyncPodUpdate, pod, nil, podStatus) // podStatus is as the previous call with one running container
if err != nil {
	t.Errorf("Unexpected err: %v", err)
}

// The metric should not consider the running pod again.
testutil.AssertHistogramTotalCount(t, "kubelet_pod_start_duration_seconds", map[string]string{}, 1)
```

I'm not sure whether table-driven testing is needed here. What I understand is that we want to test the behavior of `SyncPod` over changes in the pod's status and `statusManager` keys. If we used table-driven testing, cases would be dependent on each other, which doesn't seem really clean to me.

natasha41575 replied:
SGTM

natasha41575 replied:
It's fine with me if you prefer to do it not table-driven as long as we have the coverage.

ali-a-a replied:
Thanks for your comment. I changed the unit test to cover both `!ok && apiPodStatus.Phase == v1.PodRunning` and `existingStatus.Phase == v1.PodRunning` cases. 

I also tried table-driven testing, but since the counter of the histogram cannot be reset easily, the results of some cases would be dependent on their previous cases. So, I didn't use it.

natasha41575 replied:
> I also tried table-driven testing, but since the counter of the histogram cannot be reset easily, the results of some cases would be dependent on their previous cases. So, I didn't use it.

We have other table driven tests that have this behavior where some cases are affected by previous cases. If that's the only reason you decided not to use it, we should still go with a table-driven structure. It will be easier to read and see that we are covering all the cases.

Something like:

```
	tests := []struct {
		name                  string
		pod                   *v1.Pod
                expectedHistogramBump bool
	}{
        ...
```

natasha41575 commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -1407,6 +1407,109 @@ func TestCreateMirrorPod(t *testing.T) {
+			{
+				Name:  "container1",
+				State: kubecontainer.ContainerStateCreated,
+			},
```
why does the container status here need to be populated? Can we instead just pass in `&kubecontainer.PodStatus{ID: pod.UID}` directly into the `SyncPod` call?

natasha41575 commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -1407,6 +1407,109 @@ func TestCreateMirrorPod(t *testing.T) {
+	}
+
+	// In the first sync process, pod1 is pending.
+	_, err := kubelet.SyncPod(context.Background(), kubetypes.SyncPodSync, pod1, nil, pod1Status)
```
It looks to me like you are relying on this `SyncPod` call to populate the status manager so that we cover the `ok && existingStatus.Phase == v1.PodPending` case later. I think this is more confusing than it needs to be. Instead, each case in the table driven test should have a different pod, and whenever we need to test the case where `ok` is true, we populate the status manager directly using `kl.statusManager.SetPodStatus`. This will also reduce how much each test case depends on the previous one.

ali-a-a committed:
Use table-driven testing for TestPodStartDurationCountMetric

natasha41575 commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -1407,6 +1407,105 @@ func TestCreateMirrorPod(t *testing.T) {
+		expectedHistogramBump bool
+	}{
+		{
+			name:                  "ok is false and apiPodStatus is pending",
```
super nit: someone just reading this test won't necessarily know what `ok` means without context, let's be more explicit:

```suggestion
			name:                  "statusManager not populated and apiPodStatus is pending",
```

This also applies to the other test names that refer to `ok`

natasha41575 commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -1407,6 +1407,105 @@ func TestCreateMirrorPod(t *testing.T) {
+		},
+	}
+
+	histogramCount := 0
```
nit: is it possible to set `histogramCount` to the current total histogram count here? (you can disregard this comment if it would be too challenging or complex to do so)

natasha41575 reviewed (commented):
couple of small nits

> I think the way I wanted to test it was different than your way. I wanted to write some sort of behavioral test that works based on changes to pod status. Instead of focusing on the state of the status manager, I was focusing on transitions in the pod status. But I understand your point as well and agree with you. So, I changed the test to a table-driven test in my last commit. Please let me know if you have anything else in your mind.

Thanks for working with me on this :)

/hold cancel 

@HirazawaUi FYI that I'm removing the hold since I think the unit tests look pretty good now and from [your previous comment](https://github.com/kubernetes/kubernetes/pull/132278#issuecomment-2972841883) sounded like you were OK if we don't have a specific e2e test for this

HirazawaUi commented on pkg/kubelet/kubelet.go:
```diff
@@ -1928,7 +1928,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 	// since kubelet first saw the pod if firstSeenTime is set.
 	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
-	if !ok || existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
+	if (!ok || existingStatus.Phase == v1.PodPending) && apiPodStatus.Phase == v1.PodRunning &&
```
/cc @dashpole I'd appreciate broader and more accurate input from sig-Instrumentation on this matter.

ali-a-a replied:
> 
    Apply the following code changes:

	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
	if ok {
		if existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
			!firstSeenTime.IsZero() {
			fmt.Printf("pod %s firstSeenTime: %v", pod.Name, firstSeenTime)
			metrics.PodStartDuration.Observe(metrics.SinceInSeconds(firstSeenTime))
		}
	}

This implementation is what I wrote first, but as per talk with @natasha41575, it misses the case the pod is running right away in the first call to `SyncPod`. In that case, we would calculate the wrong duration :?  

> However, because kubelet restarted, the firstSeenTime would be approximately equal to kubelet's startup time. This would make all recorded pod startup durations appear extremely short.

Based on the current metric's description, isn't it what we want? The current description (whether right or wrong) says:

`Duration in seconds from kubelet seeing a pod for the first time to the pod starting to run`

If pods keep running and kubelet restarts, the difference between the time kubelet seeing a pod and the time the pod starts running should be almost zero as the pod is already running.

I agree that the name of this metric is really misleading, but this implementation aligns with its description. I think the fix to the name or introducing a new metric can be done in another dedicated PR.

HirazawaUi replied:
> Based on the current metric's description, isn't it what we want? The current description (whether right or wrong) says:
> 
> `Duration in seconds from kubelet seeing a pod for the first time to the pod starting to run`

This is exactly what I mentioned in https://github.com/kubernetes/kubernetes/pull/132278?notification_referrer_id=NT_kwDOAy-z_bQxNjk3MDg5NDIyMTo1MzQ1NzkxNw#discussion_r2157034088.

Our original expectation was: the `pod_start_duration_seconds metric` should record the time taken for a pod to transition from the `Pending` to the `Running` state.

Both the name `pod_start_duration_seconds` and its previous description confirm that this is what we intended.

However, we messed up in the implementation of the `pod_start_duration_seconds` metric functionality, it doesn't work as intended.

Subsequently, in #105829, we revised the description of the `pod_start_duration_seconds` metric based on this incorrect logic. This does not justify our change to the description, especially since the description now severely contradicts the metric's name.

Perhaps we need to clarify the responsibility of the `pod_start_duration_seconds` metric, revise its description again to align with the metric name, and make it function correctly.

Let's cc the sig-node maintainers, @SergeyKanzhelev @tallclair @haircommander to get broader perspectives.

natasha41575 replied:
Ah thanks @HirazawaUi for the context, I misunderstood the expectation of this metric. Based on the metric description I had assumed that re-recording the metric upon kubelet restart was intentional, but from your description I see that this was not the original desired behavior.

dashpole commented:
A few comments from a sig-instrumentation perspective (cc @kubernetes/sig-instrumentation-approvers to keep me honest):

* The `pod_start_duration_seconds` is alpha.  This means you can technically make modifications and fixes to it. But if you think a lot of users are relying on the broken behavior, you should still avoid breaking them if possible.
* Without looking into this too deeply, the metric does seem misleading to me.  If there is a way to fix it to only record the start duration for pods that pods that have started after the kubelet started, that seems like an appropriate fix to implement.

HirazawaUi commented:
> * Without looking into this too deeply, the metric does seem misleading to me.  If there is a way to fix it to only record the start duration for pods that pods that have started after the kubelet started, that seems like an appropriate fix to implement.

We can achieve this simply by applying the following change:
```
	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
	if ok {
		if existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
			!firstSeenTime.IsZero() {
			metrics.PodStartDuration.Observe(metrics.SinceInSeconds(firstSeenTime))
		}
	}
```
@ali-a-a, would you mind making this change? Or do you have other opinions?

ali-a-a commented:
> We can achieve this simply by applying the following change:
> 
> ```
> 	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
> 	if ok {
> 		if existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
> 			!firstSeenTime.IsZero() {
> 			metrics.PodStartDuration.Observe(metrics.SinceInSeconds(firstSeenTime))
> 		}
> 	}
> ```

@HirazawaUi, just to double-check. If the pod is already running in the first call to `SyncPod`, we miss its startup metric in the suggested implementation. Did you also consider this case?

HirazawaUi commented:
> just to double-check. If the pod is already running in the first call to `SyncPod`, we miss its startup metric in the suggested implementation. Did you also consider this case?

If we want to record only the start duration of pods launched after kubelet starts, this is a good solution. What other issues might it have? I haven't thought of any yet.

Aside from kubelet restarts, I don’t see any scenario in the code path where a pod is already in the running state during its first `SyncPod` execution. SyncPod drives the transition of pod states, so it observes the entire lifecycle of a pod. Please correct me if I'm wrong.

Or perhaps you're being cautious and feel this requires further discussion and more precise suggestions.  You could hold off on making this change for now.

ali-a-a commented:
> Aside from kubelet restarts, I don’t see any scenario in the code path where a pod is already in the running state during its first `SyncPod` execution. SyncPod drives the transition of pod states, so it observes the entire lifecycle of a pod. Please correct me if I'm wrong.

I hadn't looked into the process of syncing pods in depth and brought that question up to ensure we are aware of that. After looking at the code, I also couldn't find a case where that scenario would happen. I think we can go with this solution.

@natasha41575 as you had this comment (https://github.com/kubernetes/kubernetes/pull/132278#discussion_r2145394908) on this solution before, does this change sound good to you as well?

ali-a-a committed:
Fix the condition of recording the pod start duration

ali-a-a commented:
> We can achieve this simply by applying the following change:
> 
> ```
> 	existingStatus, ok := kl.statusManager.GetPodStatus(pod.UID)
> 	if ok {
> 		if existingStatus.Phase == v1.PodPending && apiPodStatus.Phase == v1.PodRunning &&
> 			!firstSeenTime.IsZero() {
> 			metrics.PodStartDuration.Observe(metrics.SinceInSeconds(firstSeenTime))
> 		}
> 	}
> ```

As there were no other ideas, I changed the code based on this suggestion. I also reverted related comments in PR https://github.com/kubernetes/kubernetes/pull/105829. Please let me know if you find any problems with the new changes.

HirazawaUi commented:
/lgtm

The changes in this PR align the implementation, comments, and naming of the `kubelet_pod_start_duration_seconds` metric for consistency. This modification has been acknowledged by sig-instrumentation: https://github.com/kubernetes/kubernetes/pull/132278#issuecomment-2996602040

However, if the kubelet restarts, pods that were started before the restart will not be included in the `kubelet_pod_start_duration_seconds` metric. I reviewed the implementation of other metrics in the kubelet, and we do not require all pods on the node to be reflected in the metrics after a kubelet restart. Therefore, this behavior remains reasonable.

For more detailed context, please refer to:
https://github.com/kubernetes/kubernetes/pull/132278#discussion_r2157003462
and
https://github.com/kubernetes/kubernetes/pull/132278#discussion_r2157034088

/cc @SergeyKanzhelev @mrunalp

haircommander commented:
/hold

based on conversation during SIG node meeting, I think there's some additional investigation needed. the SLI indicates it excludes image pull time, but I don't believe this metric is currently doing that. So we need more steps to get the metric in line with SLI

SergeyKanzhelev commented:
/lgtm cancel

dgrisonnet commented:
@haircommander I don't think this metric should be in line with its SLI equivalent.
In Kubernetes we introduced this pattern of SLI metrics in the apiserver with a clear goal in mind, to be able to define better SLOs and SLAs.

1. The end-to-end metric, apiserver_request_duration_seconds is meant to measure the complete latency of apiserver requests as seen by the clients.
2. The SLI apiserver_request_sli_duration_seconds is meant to be used to define SLO and SLA for the apiserver and as such it doesn't account for requests that are delegated to 3rd party applications such as webhooks.

IIRC, the goal was similar with the `kubelet_pod_start_sli_duration_seconds` metric, which is why it was decided to not account for external factors such image pulling. But it still very important to have an end-to-end view of what's happening, especially to single-out bad actors, so I would recommend to keep the existing behavior of this metric.

HirazawaUi commented:
/remove-lifecycle rotten

HirazawaUi commented:
/remove-lifecycle rotten
